File: app/src/main/java/org/example/S3BPlusTreeSeeker.java
package org.example;

import software.amazon.awssdk.core.ResponseInputStream;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.*;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.*;

public class S3BPlusTreeSeeker implements AutoCloseable {
    static final int PAGE_SIZE = OnDiskBPlusTree.PAGE_SIZE;
    static final int MAX_KEY_SIZE = OnDiskBPlusTree.MAX_KEY_SIZE;
    static final byte NODE_LEAF = OnDiskBPlusTree.NODE_LEAF;
    static final byte NODE_INTERNAL = OnDiskBPlusTree.NODE_INTERNAL;
    static final byte NODE_OVERFLOW = OnDiskBPlusTree.NODE_OVERFLOW;

    private final S3Client s3;
    private final String bucket;
    private final String key;
    private final long fileSize;

    public S3BPlusTreeSeeker(S3Client s3, String bucket, String key) {
        this.s3 = s3;
        this.bucket = bucket;
        this.key = key;
        this.fileSize = fetchFileSize();
    }

    private long fetchFileSize() {
        HeadObjectRequest req = HeadObjectRequest.builder().bucket(bucket).key(key).build();
        return s3.headObject(req).contentLength();
    }

    @Override
    public void close() {
        s3.close();
    }

    private ByteBuffer getPage(long pageId) {
        long offset = pageId * PAGE_SIZE;
        if (offset >= fileSize) throw new IllegalArgumentException("Page out of bounds");
        GetObjectRequest req = GetObjectRequest.builder()
                .bucket(bucket)
                .key(key)
                .range("bytes=" + offset + "-" + (offset + PAGE_SIZE - 1))
                .build();
        try (ResponseInputStream<GetObjectResponse> in = s3.getObject(req)) {
            byte[] buf = new byte[PAGE_SIZE];
            int read = in.readNBytes(buf, 0, PAGE_SIZE);
            if (read < PAGE_SIZE) throw new IOException("Short read");
            return ByteBuffer.wrap(buf);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private long getRootPageId() {
        // Root page id is stored at offset 0 (first 8 bytes)
        GetObjectRequest req = GetObjectRequest.builder()
                .bucket(bucket)
                .key(key)
                .range("bytes=0-7")
                .build();
        try (ResponseInputStream<GetObjectResponse> in = s3.getObject(req)) {
            byte[] buf = new byte[8];
            int read = in.readNBytes(buf, 0, 8);
            if (read < 8) throw new IOException("Short read");
            ByteBuffer bb = ByteBuffer.wrap(buf);
            return bb.getLong();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public Iterable<byte[]> search(byte[] key) {
        if (key.length > MAX_KEY_SIZE) throw new IllegalArgumentException("Key too large");
        return () -> new Iterator<byte[]>() {
            long[] overflowPages = findOverflow(getRootPageId(), key);
            long overflowPage = (overflowPages != null) ? overflowPages[0] : -1;
            ByteBuffer buf = (overflowPage != -1) ? getPage(overflowPage) : null;
            int valueIndex = 0;
            int valueCount = 0;
            int[] valueOffsets = null;
            boolean valid = overflowPage != -1;

            private void loadOffsets() {
                if (buf == null) return;
                valueCount = buf.getInt(1);
                valueOffsets = new int[valueCount];
                int pos = 5;
                for (int i = 0; i < valueCount; i++) {
                    valueOffsets[i] = pos;
                    buf.position(pos);
                    int len = buf.getInt();
                    pos = buf.position() + len;
                }
            }

            {
                if (valid) loadOffsets();
            }

            @Override
            public boolean hasNext() {
                while (valid) {
                    if (valueIndex < valueCount) return true;
                    long next = buf.getLong(PAGE_SIZE - 8);
                    if (next == -1) {
                        valid = false;
                        return false;
                    }
                    buf = getPage(next);
                    valueIndex = 0;
                    loadOffsets();
                }
                return false;
            }

            @Override
            public byte[] next() {
                if (!hasNext()) throw new NoSuchElementException();
                buf.position(valueOffsets[valueIndex]);
                int len = buf.getInt();
                byte[] val = new byte[len];
                buf.get(val);
                valueIndex++;
                return val;
            }
        };
    }

    // Returns [head, tail] or null if not found
    private long[] findOverflow(long pageId, byte[] key) {
        ByteBuffer buf = getPage(pageId);
        byte type = buf.get(0);

        if (type == NODE_LEAF) {
            List<OnDiskBPlusTree.LeafEntry> entries = readLeafEntries(buf);
            int idx = findKeyIndex(entries, key);
            if (idx != -1) {
                OnDiskBPlusTree.LeafEntry e = entries.get(idx);
                return new long[]{e.overflowHeadPage, e.overflowTailPage};
            }
            return null;
        }

        if (type == NODE_INTERNAL) {
            long[] leftmostChild = new long[1];
            List<OnDiskBPlusTree.InternalEntry> entries = readInternalEntries(buf, leftmostChild);
            if (entries.isEmpty()) throw new IllegalStateException("Internal node has no children");
            for (OnDiskBPlusTree.InternalEntry e : entries) {
                if (OnDiskBPlusTree.compareByteArrays(key, e.key) < 0)
                    return findOverflow(leftmostChild[0], key);
                leftmostChild[0] = e.rightChild;
            }
            return findOverflow(leftmostChild[0], key);
        }

        throw new IllegalStateException("Unknown page type");
    }

    private int findKeyIndex(List<OnDiskBPlusTree.LeafEntry> entries, byte[] key) {
        int left = 0, right = entries.size() - 1;
        while (left <= right) {
            int mid = (left + right) >>> 1;
            int cmp = OnDiskBPlusTree.compareByteArrays(entries.get(mid).key, key);
            if (cmp == 0) return mid;
            if (cmp < 0) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }

    private List<OnDiskBPlusTree.LeafEntry> readLeafEntries(ByteBuffer buf) {
        int count = buf.getInt(1);
        List<OnDiskBPlusTree.LeafEntry> entries = new ArrayList<>();
        int pos = 5;
        for (int i = 0; i < count; i++) {
            buf.position(pos);
            int klen = buf.getInt();
            byte[] key = new byte[klen];
            buf.get(key);
            long head = buf.getLong();
            long tail = buf.getLong();
            entries.add(new OnDiskBPlusTree.LeafEntry(key, head, tail));
            pos = buf.position();
        }
        return entries;
    }

    private List<OnDiskBPlusTree.InternalEntry> readInternalEntries(ByteBuffer buf, long[] leftmostChild) {
        int count = buf.getInt(1);
        int pos = 5;
        leftmostChild[0] = buf.getLong(pos);
        pos += 8;
        List<OnDiskBPlusTree.InternalEntry> entries = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            buf.position(pos);
            int klen = buf.getInt();
            byte[] key = new byte[klen];
            buf.get(key);
            long right = buf.getLong();
            entries.add(new OnDiskBPlusTree.InternalEntry(key, right));
            pos = buf.position();
        }
        return entries;
    }
}

File: python-app/proto_streamed_file/S3BPlusTreeSeeker.py
import struct
from typing import Iterator, Optional, List, Tuple

PAGE_SIZE = 1024 * 4
MAX_KEY_SIZE = 128
MAX_VALUE_SIZE = 8

NODE_LEAF = 1
NODE_INTERNAL = 2
NODE_OVERFLOW = 3

class LeafEntry:
    def __init__(self, key: bytes, overflow_head: int, overflow_tail: int):
        self.key = key
        self.overflow_head = overflow_head
        self.overflow_tail = overflow_tail

class InternalEntry:
    def __init__(self, key: bytes, right_child: int):
        self.key = key
        self.right_child = right_child

class S3BPlusTreeSeeker:
    def __init__(self, bucket: str, key: str, s3_client):
        self.bucket = bucket
        self.key = key
        self.s3 = s3_client
        self.file_size = self._get_file_size()
        self.next_page_id = self.file_size // PAGE_SIZE

    def _get_file_size(self) -> int:
        resp = self.s3.head_object(Bucket=self.bucket, Key=self.key)
        return resp['ContentLength']

    def get_page(self, page_id: int) -> memoryview:
        offset = page_id * PAGE_SIZE
        resp = self.s3.get_object(
            Bucket=self.bucket,
            Key=self.key,
            Range=f'bytes={offset}-{offset+PAGE_SIZE-1}'
        )
        data = resp['Body'].read()
        if len(data) < PAGE_SIZE:
            data = data + b'\x00' * (PAGE_SIZE - len(data))
        return memoryview(data)

    def get_root_page_id(self) -> int:
        resp = self.s3.get_object(
            Bucket=self.bucket,
            Key=self.key,
            Range='bytes=0-7'
        )
        data = resp['Body'].read()
        return struct.unpack_from('>Q', data, 0)[0]

    def search(self, key: bytes) -> Iterator[bytes]:
        if len(key) > MAX_KEY_SIZE:
            raise ValueError("Key exceeds maximum size")
        overflow_pages = self._find_overflow(self.get_root_page_id(), key)
        if not overflow_pages:
            return iter([])
        overflow_page, tail_page = overflow_pages
        return self._overflow_iterator(overflow_page)

    def _overflow_iterator(self, overflow_page: int) -> Iterator[bytes]:
        page_id = overflow_page
        while page_id != -1:
            buf = self.get_page(page_id)
            node_type = buf[0]
            assert node_type == NODE_OVERFLOW
            value_count = struct.unpack_from('>I', buf, 1)[0]
            pos = 5
            for _ in range(value_count):
                klen = struct.unpack_from('>I', buf, pos)[0]
                pos += 4
                val = bytes(buf[pos:pos+klen])
                pos += klen
                yield val
            page_id = struct.unpack_from('>q', buf, PAGE_SIZE - 8)[0]

    def _find_overflow(self, page_id: int, key: bytes) -> Optional[Tuple[int, int]]:
        buf = self.get_page(page_id)
        node_type = buf[0]
        if node_type == NODE_LEAF:
            entries = self._read_leaf_entries(buf)
            idx = self._find_key_index(entries, key)
            if idx != -1:
                entry = entries[idx]
                return (entry.overflow_head, entry.overflow_tail)
            return None
        elif node_type == NODE_INTERNAL:
            leftmost_child, entries = self._read_internal_entries(buf)
            for entry in entries:
                if key < entry.key:
                    return self._find_overflow(leftmost_child, key)
                leftmost_child = entry.right_child
            return self._find_overflow(leftmost_child, key)
        else:
            raise Exception("Unknown page type")

    def _find_key_index(self, entries: List[LeafEntry], key: bytes) -> int:
        left, right = 0, len(entries) - 1
        while left <= right:
            mid = (left + right) // 2
            cmp = (entries[mid].key > key) - (entries[mid].key < key)
            if cmp == 0:
                return mid
            elif cmp < 0:
                left = mid + 1
            else:
                right = mid - 1
        return -1

    def _read_leaf_entries(self, buf: memoryview) -> List[LeafEntry]:
        count = struct.unpack_from('>I', buf, 1)[0]
        entries = []
        pos = 5
        for _ in range(count):
            klen = struct.unpack_from('>I', buf, pos)[0]
            pos += 4
            key = bytes(buf[pos:pos+klen])
            pos += klen
            head = struct.unpack_from('>Q', buf, pos)[0]
            pos += 8
            tail = struct.unpack_from('>Q', buf, pos)[0]
            pos += 8
            entries.append(LeafEntry(key, head, tail))
        return entries

    def _read_internal_entries(self, buf: memoryview) -> Tuple[int, List[InternalEntry]]:
        count = struct.unpack_from('>I', buf, 1)[0]
        pos = 5
        leftmost_child = struct.unpack_from('>Q', buf, pos)[0]
        pos += 8
        entries = []
        for _ in range(count):
            klen = struct.unpack_from('>I', buf, pos)[0]
            pos += 4
            key = bytes(buf[pos:pos+klen])
            pos += klen
            right = struct.unpack_from('>Q', buf, pos)[0]
            pos += 8
            entries.append(InternalEntry(key, right))
        return leftmost_child, entries
